File: Labs/Lab5/main.py
class TreeNode(object):

    def __init__(self, val=0, left=None, right=None):
        self.val = val

        self.left = left

        self.right = right


# This function creates an array of nodes in
# in-order traversal order of the tree given the root
def inOrder(self, root):
    left = []
    right = []
    if root.left:
        left = inOrder(inOrder, root.left)
    if root.right:
        right = inOrder(inOrder, root.right)
    arr = []
    for node in left:
        arr.append(node)
    arr.append(root)
    for node in right:
        arr.append(node)
    return arr


# Checks to see if the BST is valid and returns a bool
def isValidBST(self, root):
    # Make a bound? optimal is 5 lines
    # is valid until proven false
    valid = True
    # gets an array of the BST in in-order traversal
    array = inOrder(inOrder, root)
    # for each node in the array it compares if the one
    # after is bigger than the current one. If it is not
    # then this is not a valid BST since it should be
    # in ascending order in in-order traversal if it is valid.
    for i in range(len(array) - 1):
        if array[i].val >= array[i+1].val:
            valid = False
    return valid


root4 = TreeNode(4)
node3 = TreeNode(3)
node2 = TreeNode(2)
node1 = TreeNode(1)
node5 = TreeNode(5)
node6 = TreeNode(6)
node7 = TreeNode(7)
node8 = TreeNode(8)
node9 = TreeNode(9)

root4.left = node3
root4.right = node8
node3.left = node1
node8.left = node7
node8.right = node9

print(isValidBST(isValidBST, root4))

File: Labs/Lab5/Main.java
import java.util.ArrayList;

public class Main {

    // Function returns ArrayList of TreeNodes in in-order
    // traversal given the root of the tree.
    public ArrayList<TreeNode> inOrder(TreeNode root) {
        ArrayList<TreeNode> left = new ArrayList<TreeNode>();
        ArrayList<TreeNode> right = new ArrayList<TreeNode>();
        if (root.left != null) {
            left = inOrder(root.left);
        }
        if (root.right != null) {
            right = inOrder(root.right);
        }
        ArrayList<TreeNode> arr = new ArrayList<TreeNode>();
        for (TreeNode node : left) {
            arr.add(node);
        }
        arr.add(root);
        for (TreeNode node : right) {
            arr.add(node);
        }
        return arr;
    }

    // Checks if the tree is a BST
    public boolean isValidBST(TreeNode root) {
        boolean valid = true;
        // array is an in-order traversal of the tree
        ArrayList<TreeNode> array = inOrder(root);

        // checks that each value is less than the next
        // if it is not this is not a BST since it should
        // be in order
        for (int i = 0; i < array.size() - 1; i++) {
            if (array.get(i).val >= array.get(i+1).val) {
                valid = false;
            }
        }
        return valid;
    }

    public static void main(String[] args) {
        TreeNode root4 = new TreeNode(4);
        TreeNode node3 = new TreeNode(3);
        TreeNode node2 = new TreeNode(2);
        TreeNode node1 = new TreeNode(1);
        TreeNode node4 = new TreeNode(4);
        TreeNode node5 = new TreeNode(5);
        TreeNode node6 = new TreeNode(6);
        TreeNode node7 = new TreeNode(7);
        TreeNode node8 = new TreeNode(8);
        TreeNode node9 = new TreeNode(9);

        root4.left = node3;
        root4.right = node8;
        node3.right = node4;
        node8.left = node7;
        node8.right = node9;

        Main obj = new Main();
        System.out.print(obj.isValidBST(root4));
    }
}